### ISSUE TITLE
Dead code deletion deleting code that's not dead

### ISSUE THREAD

**Issue by aboyton** [vimeo/psalm#9518]
Running Psalm with dead code deletion is incorrectly removing code that is sadly not dead. This appears to be incorrectly marking variables defined inside an `if` statement as dead when they are not. While the follow code isn't great code (it's hard to reason about that `$b` will always be defined when used), Psalm is breaking it.

```php
function foo(int $i) {
    if ($i === 1) {
        $b = 3;
    }
    // Do some shared logic
    echo "Hello world\n";

    if ($i === 1) {
        echo $b;
    }
}
```

is turned into
```php
function foo(int $i): void {
    $b = null;
    if ($i === 1) {
    }
    // Do some shared logic
    echo "Hello world\n";

    if ($i === 1) {
        echo $b;
    }
}
```
which does not do the same thing when called with `foo(1)`.

Repo: https://psalm.dev/r/ec292cb89b

**Comment by orklah** [comment]
Yeah, Psalm has a pretty hard time with conditionally defined variables
